uvm/bootstrap/com/untangle/uvm/UvmContextSelector.java

```diff
@@ -0,0 +1,248 @@
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Custom ContextSelector implementation to create multiple logger context for apps
+ */
+public class UvmContextSelector implements ContextSelector {
+
+    private final ConcurrentHashMap<String, LoggerContext> contexts = new ConcurrentHashMap<>();
+    private static final StatusLogger LOGGER = StatusLogger.getLogger();
+    public static final String UVM_LOG = "uvm";
+    private static final UvmContextSelector INSTANCE;
+    private static final ThreadLocal<String> THREAD_LOG_INFO = new InheritableThreadLocal<>();
+
+    static {
+        INSTANCE = new UvmContextSelector();
+    }
+
+    /**
+     * instance() Provides the UvmContextSelector singleton
+     * @return UvmContextSelector
+     */
+    public static UvmContextSelector instance() {
+        return INSTANCE;
+    }
+
+    /**
+     * Overrides getContext method of ContextSelector interface
+     * @param fqcn fqcn
+     * @param loader loader
+     * @param currentContext boolean isCurrentContext
+     * @return LoggerContext
+     */
+    @Override
+    public LoggerContext getContext(String fqcn, ClassLoader loader, boolean currentContext) {
+        try{
+            String contextName = THREAD_LOG_INFO.get();
+            synchronized (contexts) {
+                if (!contexts.containsKey(contextName)) {
+                    UvmLoggerContext context = new UvmLoggerContext(contextName);
+                    context.start();
+                    contexts.put(contextName, context);
+                }
+            }
+            return contexts.get(contextName);
+        } catch (Exception e) {
+            LOGGER.error("Exception: ", e);
+        }
+        return contexts.get(UVM_LOG);
+    }
+
+    /**
+     * Overrides getContext method of ContextSelector interface
+     * @param fqcn fqcn
+     * @param loader loader
+     * @param currentContext boolean isCurrentContext
+     * @param configLocation configLocation
+     * @return LoggerContext
+     */
+    @Override
+    public LoggerContext getContext(String fqcn, ClassLoader loader, boolean currentContext, URI configLocation) {
+        return getContext(fqcn, loader, currentContext);
+    }
+
+    /**
+     * Overrides removeContext method of ContextSelector interface
+     * @param context LoggerContext
+     */
+    @Override
+    public void removeContext(LoggerContext context) {
+        synchronized (contexts) {
+            LoggerContext remove = contexts.remove(context.getName());
+        }
+    }
+
+    /**
+     * Overrides getLoggerContexts method of ContextSelector interface
+     * @return List<LoggerContext>
+     */
+    @Override
+    public List<LoggerContext> getLoggerContexts() {
+        Collection<LoggerContext> values = contexts.values();
+        return new ArrayList<>(values);
```

Comment:
How about just,
```
return new ArrayList<>(contexts.values());
```
Also, don't you need to surround return statement with `synchronised(contexts)` here ?

Reply:
Changed to `return new ArrayList<>(contexts.values());`

As we are not modifying contexts in this method i don't think we need to use synchronised blcok.